core/trino-main/src/main/java/io/trino/sql/planner/ArraySortAfterArrayDistinctRewriter.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.trino.sql.planner;
+
+import com.google.common.collect.ImmutableMap;
+import io.trino.Session;
+import io.trino.metadata.Metadata;
+import io.trino.operator.scalar.ArrayDistinctFunction;
+import io.trino.operator.scalar.ArraySortFunction;
+import io.trino.spi.type.Type;
+import io.trino.sql.tree.Expression;
+import io.trino.sql.tree.ExpressionRewriter;
+import io.trino.sql.tree.ExpressionTreeRewriter;
+import io.trino.sql.tree.FunctionCall;
+import io.trino.sql.tree.NodeRef;
+import io.trino.sql.tree.QualifiedName;
+import io.trino.sql.tree.SymbolReference;
+
+import java.util.List;
+import java.util.Map;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+public final class ArraySortAfterArrayDistinctRewriter
+{
+    public static Expression rewrite(Expression expression, Map<NodeRef<Expression>, Type> expressionTypes, Metadata metadata)
+    {
+        return ExpressionTreeRewriter.rewriteWith(new Visitor(expressionTypes, metadata), expression);
+    }
+
+    private ArraySortAfterArrayDistinctRewriter() {}
+
+    public static Expression rewrite(Expression expression, Session session, Metadata metadata, TypeAnalyzer typeAnalyzer, TypeProvider typeProvider)
+    {
+        requireNonNull(metadata, "metadata is null");
+        requireNonNull(typeAnalyzer, "typeAnalyzer is null");
+
+        if (expression instanceof SymbolReference) {
+            return expression;
+        }
+        Map<NodeRef<Expression>, Type> expressionTypes = typeAnalyzer.getTypes(session, typeProvider, expression);
+
+        return rewrite(expression, expressionTypes, metadata);
+    }
+
+    private static class Visitor
+            extends ExpressionRewriter<Void>
+    {
+        private final Map<NodeRef<Expression>, Type> expressionTypes;
+        private final Metadata metadata;
+
+        public Visitor(Map<NodeRef<Expression>, Type> expressionTypes, Metadata metadata)
+        {
+            this.expressionTypes = ImmutableMap.copyOf(requireNonNull(expressionTypes, "expressionTypes is null"));
+            this.metadata = metadata;
+        }
+
+        @Override
+        public Expression rewriteFunctionCall(FunctionCall node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            if (node.getName().getSuffix().contains(String.format("@%s@", ArrayDistinctFunction.NAME))
```

Comment:
The content of `FunctionCall.getName()` is an opaque identifier for the underlying function resolved during analysis. Attempting to inspect that string is not the right way to do this. Instead, you need to get the name from the corresponding `ResolvedFunction` object, which you can obtain from `MetadataManager` (see how it's done in `DeterminismEvaluator`):

```
metadata.decodeFunction(node.getName())
```

Ideally, you'd just compare the id of the ResolvedFunction with the id of a well-known function, like it's done in `RewriteSpatialPartitioningAggregation.isStEnvelopeFunctionCall()` but that only works if you know the specific argument types in advance. So, checking the name should be ok for now.

Reply:
Makes total sense. Will take care of it!